### PR TITLE
codegen: Improve array default handing

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45164 torch.fft: Two dimensional FFT functions
* **#45163 codegen: Improve array default handing**

Differential Revision: [D24132279](https://our.internmc.facebook.com/intern/diff/D24132279)

### CODE REVIEW

File: tools/codegen/api/cpp.py [hunk 1]
```diff
@@ -153,14 +153,23 @@ def returns_type(rs: Sequence[Return]) -> str:
...
         return '{}'
+
+    if isinstance(t, OptionalType):
+        if d == 'None':
+            return 'c10::nullopt'
+
+        return default_expr(d, t.elem)
+
+    if isinstance(t, ListType) and d.startswith('[') and d.endswith(']'):
+        return '{' + d[1:-1] + '}'
```
Comment (peterbell10) [inline tools/codegen/api/cpp.py:171]: This is needed to handle `dim=[-2,-1]` in #45164 and more generally, to support arbitrary list defaults. Without it, it tries to literally put square-brackets into the C++ defaults.

**Comment by mruberry** [comment]
Going to let @ezyang review this one. He's been working on codegen recently.

File: tools/codegen/api/cpp.py (hunk 1 above)
Comment (ezyang) [reply tools/codegen/api/cpp.py:171]: I'd prefer it if we can make this fail noisily if someone accidentally, e.g., prefixes the default value with an extra space. Can we do something like, assert `d.startswith('[') and d.endswith(']')` when `isinstance(t, ListType)`? That would ensure that the person specified braces.

**Review by ezyang — APPROVED** [review]

File: tools/codegen/api/cpp.py (hunk 1 above)
Comment (peterbell10) [reply tools/codegen/api/cpp.py]: That check was there to allow sized lists with scalar defaults e.g. `int[2] padding=0`. So, I've put in an exception for unsized lists only.

**Comment by mruberry** [comment]
Sorry to bother you @peterbell10, but would you rebase this please? Our internal tools are struggling to merge it properly.

**Comment by peterbell10** [comment]
@mruberry I fixed the merge conflict.

**Pull request closed without merging**